src/websocket.c

```diff
@@ -142,57 +196,76 @@ ws_handshake_reply(struct http_client *c) {
 
 	/* Concat all */
 
-	/* template0 */
-	memcpy(p, template0, sizeof(template0)-1);
-	p += sizeof(template0)-1;
-	memcpy(p, origin, origin_sz);
-	p += origin_sz;
+	/* template_start */
+	memcpy(p, template_start, sizeof(template_start)-1);
+	p += sizeof(template_start)-1;
+
+	/* template_accept */
+	memcpy(p, template_accept, sizeof(template_accept)-1);
+	p += sizeof(template_accept)-1;
+	memcpy(p, &sha1_handshake[0], handshake_sz);
+	p += handshake_sz;
+
+	/* template_sec_origin */
+	if(origin && origin_sz) {
+		memcpy(p, template_sec_origin, sizeof(template_sec_origin)-1);
+		p += sizeof(template_sec_origin)-1;
+		memcpy(p, origin, origin_sz);
+		p += origin_sz;
+	}
 
-	/* template1 */
-	memcpy(p, template1, sizeof(template1)-1);
-	p += sizeof(template1)-1;
+	/* template_loc */
+	memcpy(p, template_loc, sizeof(template_loc)-1);
+	p += sizeof(template_loc)-1;
 	memcpy(p, host, host_sz);
 	p += host_sz;
 	memcpy(p, c->path, c->path_sz);
 	p += c->path_sz;
 
-	/* template2 */
-	memcpy(p, template2, sizeof(template2)-1);
-	p += sizeof(template2)-1;
-	memcpy(p, host, host_sz);
-	p += host_sz;
+	/* template_end */
+	memcpy(p, template_end, sizeof(template_end)-1);
+	p += sizeof(template_end)-1;
 
-	/* template3 */
-	memcpy(p, template3, sizeof(template3)-1);
-	p += sizeof(template3)-1;
-	memcpy(p, &sha1_handshake[0], handshake_sz);
-	p += handshake_sz;
-
-	/* template4 */
-	memcpy(p, template4, sizeof(template4)-1);
-	p += sizeof(template4)-1;
-
-	/* build HTTP response object by hand, since we have the full response already */
-	struct http_response *r = calloc(1, sizeof(struct http_response));
-	if(!r) {
-		slog(c->s, WEBDIS_ERROR, "Failed to allocate response for WS handshake", 0);
-		free(buffer);
+	int add_ret = evbuffer_add(ws->wbuf, buffer, sz);
+	free(buffer);
+	if(add_ret < 0) {
+		slog(c->s, WEBDIS_ERROR, "Failed to add response for WS handshake", 0);
 		return -1;
 	}
-	r->w = c->w;
-	r->keep_alive = 1;
-	r->out = buffer;
-	r->out_sz = sz;
-	r->sent = 0;
-	http_schedule_write(c->fd, r); /* will free buffer and response once sent */
+
+	ws_schedule_write(ws); /* will free buffer and response once sent */
 
 	return 0;
 }
 
+static void
+ws_log_cmd(struct ws_client *ws, struct cmd *cmd) {
+	char log_msg[SLOG_MSG_MAX_LEN];
+	char *p = log_msg, *eom = log_msg + sizeof(log_msg) - 1;
+	if(!slog_enabled(ws->http_client->s, WEBDIS_DEBUG)) {
+		return;
+	}
+
+	memset(log_msg, 0, sizeof(log_msg));
+	memcpy(p, "WS: ", 4); /* WS prefix */
+	p += 4;
```

Comment:
since this is making assumptions about `SLOG_MSG_MAX_LEN` there could be a compile-time check in `slog.h` that it's at least 4 bytes + enough for the date/time format, etc.